### ISSUE TITLE
Issue with ngc JS output for @Injectable when using providedIn: 'root' and tsconfig paths

### ISSUE THREAD

**Issue by bfarrelly** [angular/angular#24773]
**Note:** I have also raised this in the Angular-CLI repo as I am not sure which one is the most appropriate place to discuss **ngc** or if the issues may also impact ng.

### Bug Report or Feature Request (mark with an `x`)
```
- [x] bug report -> please search issues before submitting
- [ ] feature request
```

### Area
```
- [x] devkit
- [ ] schematics
```

### Versions

Angular v6.0.7
node v8.9.4
npm v6.1.0
win 10

### Repro steps

Firstly, I am not using **ng**, I am using **ngc** as we use SystemJS and Rollup due to the extra flexibility they give us (We have a rather large in house enterprise platform we are transitioning from AngularJS that makes extensive use of lazy loading).

The steps:

- Create two separate libraries, with a service each (Service A and Service B) using @Injectable with providedIn: 'root'
- Export serviceB using an index.js for use in serviceA
- npm link Lib B to Lib A
- Create a path mapping in lib A's tsconfig to lib B module in tsconfig e.g. 

    `"lib-two": ["../node_modules/lib-two/build/src"]`

- Import serviceB into serviceA.

ServiceA.ts

```
import { Injectable } from "@angular/core";
import { ServiceB } from 'lib-two';

@Injectable({
    providedIn: 'root'
})
export class ServiceA {
    constructor(private serviceB: ServiceB) {

    }
}
```
ServiceB.ts
```
import { Injectable } from "@angular/core";

@Injectable({
    providedIn: 'root'
})
export class ServiceB {
    constructor() {

    }
}
```

- Compile using ngc, target: es5, module: es2015
- Observe that as part of the code gen to add the factory for @Injectable, the original imports are duplicated using import * syntax

**update:** Most importantly the path used for the namespace import of lib-two has changed and is now relative to the node module and no longer using the path defined in the tsconfig.  This breaks downstream Rollup/SystemJS workflows as the namespace import looks like a separate dependency.

ServiceA.js

```
import { Injectable } from "@angular/core";
import { ServiceB } from 'lib-two';
import * as i0 from "@angular/core";
import * as i1 from "lib-two/build/src/index";
var ServiceA = /** @class */ (function () {
    function ServiceA(serviceB) {
        this.serviceB = serviceB;
    }
    ServiceA.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ServiceA.ctorParameters = function () { return [
        { type: ServiceB }
    ]; };
    ServiceA.ngInjectableDef = i0.defineInjectable({ factory: function ServiceA_Factory() { return new ServiceA(i0.inject(i1.ServiceB)); }, token: ServiceA, providedIn: "root" });
    return ServiceA;
}());
export { ServiceA };
```

ServiceB.js

```
import { Injectable } from "@angular/core";
import * as i0 from "@angular/core";
var ServiceB = /** @class */ (function () {
    function ServiceB() {
    }
    ServiceB.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ServiceB.ctorParameters = function () { return []; };
    ServiceB.ngInjectableDef = i0.defineInjectable({ factory: function ServiceB_Factory() { return new ServiceB(); }, token: ServiceB, providedIn: "root" });
    return ServiceB;
}());
export { ServiceB };
```
This isn't what I expected.  Of course when I go to use rollup to turn this into a systemJS module it adds the additional imports, breaking my SystemJS mappings.

I did think that I had a workaround by adding my own factory using the _useFactory_ attribute:

ServiceA.ts
```
import { Injectable, inject } from "@angular/core";
import { ServiceB } from 'lib-two';

@Injectable({
    providedIn: 'root',
    useFactory: () => {
        return new ServiceA(inject(ServiceB))
    }
})
export class ServiceA {
    constructor(private serviceB: ServiceB) {

    }
}
```

Which then resulted in some workable code.

ServiceA.js
```
import { Injectable, inject } from "@angular/core";
import { ServiceB } from 'lib-two';
import * as i0 from "@angular/core";
var ɵ0 = function () {
    return new ServiceA(inject(ServiceB));
};
var ServiceA = /** @class */ (function () {
    function ServiceA(serviceB) {
        this.serviceB = serviceB;
    }
    ServiceA.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root',
                    useFactory: ɵ0
                },] }
    ];
    /** @nocollapse */
    ServiceA.ctorParameters = function () { return [
        { type: ServiceB }
    ]; };
    ServiceA.ngInjectableDef = i0.defineInjectable({ factory: ɵ0, token: ServiceA, providedIn: "root" });
    return ServiceA;
}());
export { ServiceA };
export { ɵ0 };
```
This worked for a few days however today I hit another issue (in a more complex example) in which a successful compile resulted in the code generation referencing a type from index.js '../../index' rather than relative path './ServiceC', causing cyclic dependency warnings to start showing up when bundling with rollup. 

Generated code
```
import { Injectable } from '@angular/core';
import { PluginStarter } from './plugin-starter-service';
import _ from 'lodash';
import * as i0 from "@angular/core";
import * as i1 from "../../index";

....

PluginRegistry.ngInjectableDef = i0.defineInjectable(
    { factory: ɵ0, token: i1.PluginRegistry, providedIn: "root" });
```
So time to understand/fix the root cause...

### The log given by the failure

No logs... just broken code

### Desired functionality

I don't understand why the import statements and types need to reproduced to support the generation of the factory? Why cant it continue to use the types provided?  Is there a better way that we could be doing this and still get the same outcome?

**core issue:** why is the path being rewritten without respecting the path given in the original import?

We need to use paths as we have npm packages with multiple submodules inside them.  We use paths to point the compiler at the right place and ensure that the import "symbol" remains consistent to our convention for SystemJS mapping/loading.

Coming from an AngularJS background, we also make heavy use of global services in all of our existing app content, so this feature was the one that made us decide it was finally time to look seriously at making the move.

I am just starting out on my "Angular" journey, so this may all be due to a lack of knowledge, but I figure I have to ask as I cant see a way forward for our platform at the moment.

Hope someone can help!

### Mention any other details that might be useful

The problem goes away when you remove "providedIn" as it  doesn't do the factory code generation, it also doesn't inject any services!

Here is the example in git.  I have checked in the compiled source as well just in case you don't want to pull it down and build it yourself.

[Github - Lib One](https://github.com/bfarrelly/angular-injectable-issue-lib-one)
[Github - Lib Two](https://github.com/bfarrelly/angular-injectable-issue-lib-two)

This has been driving me nuts for a few days now so any help/guidance would be appreciated!

Thanks.

**Comment by bfarrelly** [comment]
@trotyl Why the thumbs down? A bit of guidance would be useful.

**Comment by DeepamJyotiDutta123** [comment]
Why the thumbs are down??

**Comment by trotyl** [comment]
Multiple imports for same module is valid and expected, if you have any question for your build tools please raise tickets there.

**Comment by bfarrelly** [comment]
@trotyl thanks for taking the time to explain.  My question is, why is it changing the path? I am fine with it having it’s own copy to work with, I just expect it to use the path provided.  Is this unreasonable?

The duplication in the @angular context is fine as the paths match so they are resolved to the same dependency which doesn’t cause any issues.  The problem I am experiencing is caused by the compiler not respecting the import path given for the non angular dependency and putting in its own path resolved to the node modules folder, ultimately looking like a separate dependency.

Appreciate your time so far.

**Comment by trotyl** [comment]
Thanks for the explanation, understood now.

**Comment by bfarrelly** [comment]
Cheers trotyl, I have added a little detail to my initial report to try to make my ask a little clearer on first read

**Comment by bfarrelly** [comment]
Update on this one...

- Creating my own factory is still working as a workaround for the injected types
- I have pushed forward ignoring the cyclic dependency error coming from rollup (as annoying as it is) as the code still seems to work fine.

Both issues seem to occur due to the compiler not respecting provided paths, so at a guess I would say it has something to do with its resolution logic not matching typescripts.

One interesting thing is that if the type you are trying to construct is not being reexported in the barrel, then it will happily use the relative path provided.  As soon as it is reexported then is will reference the barrel rather than the relative path, causing the circular dependency in the generated code.

A permanent solution/guidance on a better way of managing inter library dependencies would certainly be useful though as relying on hacks will get old very quick!

**Comment by robwormald** [comment]
@bfarrelly any chance you can provide a simple reproduction of this? 

One thing to check - are you (like the Angular modules do) generating the entry point file? This is a bit wordy, but basically, if you check out a package like `@angular/elements` - you'll see there's a `public_api.ts` file, and *thats* the one we point ngc at - see https://github.com/angular/angular/blob/master/packages/elements/tsconfig-build.json#L24-L25

In general, Angular will use the *metadata.json* files (rather than scanning the TS/d.ts) sources across packages - setting up the metadata as above will instruct the compiler to use that import path, rather than whatever it resolves.

This is transparent if you're using the CLI to produce libraries, but if you're using `ngc` you'll need to set it up manually...

**Comment by bfarrelly** [comment]
@robwormald thanks for the response.  I have actually added the repro above, but I understand how it may have been lost in the commentary.  Please find the exact steps below.

I don't have a `public_api.ts` file but I do have an `index.ts` file.  Is it worth me trying to add one?  I am not quite sure how it differs from index.ts (as I would be reexporting it anyway) and how it may solve the cyclic dependency problem.

I have also been generating the metadata files, and interestingly enough the `service-a.service.metadata.json` file generated in the repro below still has the module defined correctly.

The cyclic dependency repro will take me some time to put together however in essence it seems to be when a service imports another service using relative path the compiler will drop back to the barrel, causing a circular dependency, which makes sense as:

_`index.js` exports `service-a.service.js` which then imports `service-b.service.js` from `index.js`_

**Repro steps**

```
git clone git@github.com:bfarrelly/angular-injectable-issue-lib-one.git

git clone 
git@github.com:bfarrelly/angular-injectable-issue-lib-two.git

cd angular-injectable-issue-lib-two

npm install

npm run build

npm link

cd ..

cd angular-injectable-issue-lib-one

npm link lib-two

npm install

npm build

```
You should be able to see the compiled service in the build folder.

`angular-injectable-issue-lib-one/build/src/services/service-a.service.js`

The path for lib-two has changed.

Import path: `lib-two` (as per tsconfig)
Generated path: `lib-two/build/src/index`

Please let me know if you need anymore info.

Cheers,

Ben

**Comment by alxhub** [comment]
This is fixed in v9 with Ivy!
